### PR TITLE
Port CodeGenPrepare to new pass manager (and BasicBlockSectionsProfil…

### LINKED ISSUE #64560: Port CodeGenPrepare to new pass manager

### PR DESCRIPTION
**Pull request by nickleus27**
Port CodeGenPrepare to new pass manager and dependency BasicBlockSectionsProfileReader
Fixes: #64560

### CODE REVIEW

**Comment by nickleus27** [comment]
I am new to llvm-project and would appreciate any suggestions, especially on what is needed to port `BasicBlockSectionsProfileReader` to `BasicBlockSectionsProfileReaderAnalysis`. Currently, I am reading the docs at the bottom of [Contributing](https://llvm.org/docs/Contributing.html) to understand how the `.ll` tests work and how to debug failing tests. I pushed this `Draft PR` in case anyone has suggestions.

**Comment by nickleus27** [comment]
@arsenm After debugging, it looks like I am failing some tests because `PSI` is nullptr, `PSI = MAMProxy.getCachedResult<ProfileSummaryAnalysis>(*F.getParent());` This would mean `ProfileSummaryAnalysis` is not cached and hasn't been run. It seems `CodeGenPrepare` expects PSI not to be nullptr. Any suggestions on how to make sure ProfileSummaryAnalysis is cached before getting to CodeGenPrepare? Thank you for your time.

**Comment by nickleus27** [comment]
I am thinking what I need is to require ProfileSummaryAnalysis like this: `MPM.addPass(RequireAnalysisPass<ProfileSummaryAnalysis, Module>());` which I could do in function `llvm::runPassPipeline` which is in `NewPMDriver.cpp`, but doing so I believe would make other tests potentially fail. Is there a way I could conditionally require ProfileSummaryAnalysis, or where would the right place to require the ProfileSummaryAnalysis Pass? Looking for suggestions, because I am feeling a little stuck at the moment.

**Comment by nickleus27** [comment]
Ping

**Comment by arsenm** [comment]
>  it looks like I am failing some tests because PSI is nullptr, PSI = MAMProxy.getCachedResult<ProfileSummaryAnalysis>(*F.getParent()); This would mean ProfileSummaryAnalysis is not cached and hasn't been run

getCachedResult is for optional analyses; if it's a hard requirement you should just use getResult instead to ensure it's available

File: llvm/lib/Passes/PassBuilder.cpp
Comment (arsenm) [inline llvm/lib/Passes/PassBuilder.cpp:1970]: Should avoid pre-adding dependent passes based on the name, does getResult work?

Comment (nickleus27) [reply llvm/lib/Passes/PassBuilder.cpp:1970]: No, I don't think getResult will work. I think PSI needs to be pre-cached using `RequireAnalysisPass<ProfileSummaryAnalysis, Module>()` in the pipeline, because we cannot request a result from an outer-scope analysis. I was having a hard time finding the right place to conditionally require ProfileSummaryAnalysis in the pipeline, so I just used this hack here. Any suggestions would be appreciated.

Comment (arsenm) [reply llvm/lib/Passes/PassBuilder.cpp:1970]: I vaguely recall seeing this hack in another patch by @paperchalice. For now I would drop the FirstName check and just unconditionally add it?

Comment (nickleus27) [reply llvm/lib/Passes/PassBuilder.cpp]: Thank you for the advice, the suggestions really helped.

**Review by arsenm — APPROVED** [review]
It's progress although the ProfileSummaryInfo hack is unfortunate

**Comment by nickleus27** [comment]
> It's progress although the ProfileSummaryInfo hack is unfortunate

I was able to remove the ProfileSmmaryInfo hack in the last commit, and instead require ProfileSummary in the .ll regression tests via paperchalice's example.

**Comment by arsenm** [comment]
> > It's progress although the ProfileSummaryInfo hack is unfortunate
> 
> I was able to remove the ProfileSmmaryInfo hack in the last commit, and instead require ProfileSummary in the .ll regression tests via paperchalice's example.

Requiring manual specification of it isn't great, it just moves the workaround out of the sources and into the tests

**Review by arsenm — APPROVED** [review]

**Comment by nickleus27** [comment]
@arsenm Should I squash these fixup commits? Or leave for now?

**Comment by arsenm** [comment]
Squash and merge is the only enabled option for the project

**Comment by nickleus27** [comment]
My apologies, I don't see the merge button. This is my first pull request on this project, where do I find the merge button?

**Comment by arsenm** [comment]
> My apologies, I don't see the merge button. This is my first pull request on this project, where do I find the merge button?

I guess if you aren't a project member you may not see it? I still find it odd a committer still needs to hit the button after approval

**Comment by nickleus27** [comment]
> I guess if you aren't a project member you may not see it? I still find it odd a committer still needs to hit the button after approval

I agree, and thank you.

**Comment by RKSimon** [comment]
@paperchalice @nickleus27 I've reverted this (and #77054) as they were breaking expensive checks builds: https://lab.llvm.org/buildbot/#/builders/104 - please can you take a look?

**Comment by nickleus27** [comment]
Thank you, I will work on fixing this.

**Comment by nickleus27** [comment]
I cannot reproduce the fails on my local machine (MacOS 14.2.1 x86_64). I noticed the failed tests from my commit [https://lab.llvm.org/buildbot/#/builders/16/builds/59614](https://lab.llvm.org/buildbot/#/builders/16/builds/59614) and from [https://lab.llvm.org/buildbot/#/builders/104/builds/14911](https://lab.llvm.org/buildbot/#/builders/104/builds/14911) are both Linux (Debian/Ubuntu). Could I reproduce and test a fix in a Debian/Ubuntu Virtual Machine locally? Any suggestions would be appreciated.

**Comment by boomanaiden154** [comment]
It's an expensive checks failure, so you'll need to enable expensive checks in order to reproduce it. `-DLLVM_ENABLE_EXPENSIVE_CHECKS=ON` on your CMake invocation should set it up. You'll probably also want to set `-DLLVM_OPTIMIZED_TABLEGEN=ON` as well or some of the backends will take forever to build (particularly AMDGPU).

**Comment by nickleus27** [comment]
Awesome, thank you!

**Pull request merged**
